### PR TITLE
OTG test - RT-1.2: BGP Policy & Route Installation

### PR DESCRIPTION
**Pull request by octpetre**
Hello,

This is the first PR for an OTG test into openconfig/featureprofiles. The source branch is up to date with the upstream repo.
A library called "otgutils" was added under ./internal which includes otg related stats and functions
The tests are using the latest integration of OTG in ondatra and ygot commits.

Thank you,

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,11 @@
 *~
-topologies/kne/testbed.kne.yml
 .vscode/
 # used by `make validate_paths`
-openconfig_public/
\ No newline at end of file
+openconfig_public/
+launch.json
```
Comment (liulk) [inline .gitignore:5]: Aren't launch.json and settings.json already covered by .vscode/?

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,11 @@
...
+settings.json
+topologies/kne/*.yml
+logs/
+topologies/kne/kubecfg
+feature/experimental/example_test
```
Comment (liulk) [inline .gitignore:10]: Remove?

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,11 @@
 *~
-topologies/kne/testbed.kne.yml
 .vscode/
 # used by `make validate_paths`
-openconfig_public/
\ No newline at end of file
+openconfig_public/
+launch.json
+settings.json
+topologies/kne/*.yml
```
Comment (liulk) [inline .gitignore:7]: Revert to topologies/kne/testbed.kne.yml?

File: .gitignore [hunk 4]
```diff
@@ -1,5 +1,11 @@
...
+logs/
+topologies/kne/kubecfg
```
Comment (liulk) [inline .gitignore:9]: Who created this file?

File: topologies/kne/init_arista.config [hunk 5]
```diff
@@ -0,0 +1,39 @@
+transceiver qsfp default-mode 4x10G
```
Comment (liulk) [inline topologies/kne/init_arista.config:1]: Can we merge this with arista_ceos.config?

https://github.com/openconfig/featureprofiles/blob/main/topologies/kne/arista_ceos.config

I think we can rename the result to just "arista.config"

@bstoll what do you think?

File: topologies/kne/init_arista_gribi.config [hunk 6]
```diff
@@ -0,0 +1,44 @@
+transceiver qsfp default-mode 4x10G
```
Comment (liulk) [inline topologies/kne/init_arista_gribi.config:1]: Can we just fold the gribi config into the one arista config?

File: topologies/kne/ixia-arista-3ports.textproto [hunk 7]
```diff
@@ -0,0 +1,83 @@
+name: "ixia-c"
```
Comment (liulk) [inline topologies/kne/ixia-arista-3ports.textproto:1]: Let's drop the 3ports variant and just keep the 4ports one?

File: topologies/kne/ixia-arista-ixia.textproto [hunk 8]
```diff
@@ -0,0 +1,70 @@
+name: "ixia-c"
```
Comment (liulk) [inline topologies/kne/ixia-arista-ixia.textproto:1]: Drop the 2 port variant and keep the 4 port one?

File: topologies/kne/ixia-arista-4ports.textproto [hunk 9]
```diff
@@ -0,0 +1,89 @@
+name: "ixia-c"
```
Comment (liulk) [inline topologies/kne/ixia-arista-4ports.textproto:1]: I think the name "arista_ixia.textproto" is more consistent with the existing "arista_ceos.textproto". @bstoll wdyt?

File: topologies/kne/ixia-cisco-ixia.textproto [hunk 10]
```diff
@@ -0,0 +1,81 @@
+name: "ixia-c-2"
```
Comment (liulk) [inline topologies/kne/ixia-cisco-ixia.textproto:1]: Rename to cisco_ixia.textproto and make this 4 port?

File: topologies/kne/testbed.kne.yml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+username: admin
```
Comment (liulk) [inline topologies/kne/testbed.kne.yml:1]: Remove?

File: topologies/topology_test/otg_test/topology_test.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
+// Copyright 2022 Google LLC
```
Comment (liulk) [inline topologies/topology_test/otg_test/topology_test.go:1]: Move to new path?

topologies/otg_tests/topology_test/topology_test.go

**Review by liulk — CHANGES_REQUESTED** [review]
Some early comments. Will continue the review tomorrow.

File: .gitignore (hunk 1 above)
Comment (octpetre) [reply .gitignore:5]: ...removed both....json files are part of .vscode/

File: .gitignore (hunk 2 above)
Comment (octpetre) [reply .gitignore:10]: Yes, all removed

File: .gitignore (hunk 3 above)
Comment (octpetre) [reply .gitignore:7]: I think so...initially I had multiple files based on the test, but now we can have just one I think

File: topologies/topology_test/otg_test/topology_test.go (hunk 12 above)
Comment (octpetre) [reply topologies/topology_test/otg_test/topology_test.go:1]: The original topologies/topology_test/topology_test.go was built for ate (IxNetwork) so I thought I could just put the OTG version inside of that otg_test folder. We should probably create topologies/topology_test/ate_test/ folder and move topology_test.go there.

File: topologies/kne/ixia-arista-3ports.textproto (hunk 7 above)
Comment (octpetre) [reply topologies/kne/ixia-arista-3ports.textproto:1]: Agreed

File: topologies/kne/ixia-arista-ixia.textproto (hunk 8 above)
Comment (octpetre) [reply topologies/kne/ixia-arista-ixia.textproto:1]: We could...although the 4 ports does require more resources...

File: topologies/kne/ixia-arista-4ports.textproto (hunk 9 above)
Comment (octpetre) [reply topologies/kne/ixia-arista-4ports.textproto:1]: Sounds good to me

File: topologies/kne/ixia-cisco-ixia.textproto (hunk 10 above)
Comment (octpetre) [reply topologies/kne/ixia-cisco-ixia.textproto:1]: Yes sure.

File: topologies/kne/init_arista_gribi.config (hunk 6 above)
Comment (octpetre) [reply topologies/kne/init_arista_gribi.config:1]: Yes, sure

File: topologies/kne/init_arista.config (hunk 5 above)
Comment (octpetre) [reply topologies/kne/init_arista.config:1]: I was able to successfully run the test with arista_ceos.config so I will replace init_arista.config with this

File: topologies/kne/testbed.kne.yml (hunk 11 above)
Comment (octpetre) [reply topologies/kne/testbed.kne.yml:1]: I'm not sure what exactly...

File: topologies/kne/init_arista.config (hunk 5 above)
Comment (liulk) [reply topologies/kne/init_arista.config:1]: Is this file still needed in the PR?

File: topologies/kne/init_cisco.config [hunk 13]
```diff
@@ -0,0 +1,21 @@
+!! IOS XR Configuration 7.4.1
```
Comment (liulk) [inline topologies/kne/init_cisco.config:1]: Rename this file to just cisco.config?

**Review by liulk — CHANGES_REQUESTED** [review]

File: topologies/kne/init_arista.config (hunk 5 above)
Comment (octpetre) [reply topologies/kne/init_arista.config:1]: The only difference is this line "username admin privilege 15 role network-admin nopassword" which I believe allows ssh connection without password. It's ok, I can remove the file.

File: topologies/kne/testbed.kne.yml (hunk 11 above)
Comment (octpetre) [reply topologies/kne/testbed.kne.yml:1]: I see, I removed the file now

File: topologies/topology_test/otg_test/topology_test.go (hunk 12 above)
Comment (octpetre) [reply topologies/topology_test/otg_test/topology_test.go:1]: I will update the path on both.

File: topologies/kne/init_cisco.config (hunk 13 above)
Comment (octpetre) [reply topologies/kne/init_cisco.config:1]: Done

File: go.mod [hunk 14]
```diff
@@ -1,10 +1,11 @@
 module github.com/openconfig/featureprofiles
 
-go 1.17
+go 1.18
```
Comment (liulk) [inline go.mod:3]: Ondatra is still on go 1.17. Can we keep it on 1.17 for now?

File: topologies/atedut_2.testbed [hunk 15]
```diff
@@ -1,11 +1,6 @@
-# proto-file: github.com/openconfig/ondatra/blob/main/proto/testbed.proto
-# proto-message: ondatra.Testbed
-
-# This testbed provides 2 links between a DUT and an ATE connected
-# pairwise.
-
 duts {
   id: "dut"
+  vendor: ARISTA
```
Comment (liulk) [inline topologies/atedut_2.testbed:3]: Please revert all changes in this file?

File: topologies/atedut_3.testbed [hunk 16]
```diff
@@ -0,0 +1,40 @@
+duts {
```
Comment (liulk) [inline topologies/atedut_3.testbed:1]: Tests that need 3 ports should just use the atedut_4.testbed. Delete this file?

File: topologies/topology_test/otg_tests/topology_test.go [hunk 17]
```diff
@@ -0,0 +1,194 @@
+// Copyright 2022 Google LLC
```
Comment (liulk) [inline topologies/topology_test/otg_tests/topology_test.go:1]: Sorry if I was not clear before, but the convention for test paths should be:

* ATE: topologies/ate_tests/topology_test/topology_test.go
* OTG: topologies/otg_tests/topology_test/topology_test.go

**Review by liulk — CHANGES_REQUESTED** [review]

File: topologies/atedut_3.testbed (hunk 16 above)
Comment (octpetre) [reply topologies/atedut_3.testbed:1]: Yes, removed for now. We may have to revisit this later

File: topologies/topology_test/otg_tests/topology_test.go (hunk 17 above)
Comment (octpetre) [reply topologies/topology_test/otg_tests/topology_test.go:1]: Done.

File: topologies/atedut_2.testbed (hunk 15 above)
Comment (octpetre) [reply topologies/atedut_2.testbed:3]: Done

File: go.mod (hunk 14 above)
Comment (octpetre) [reply go.mod:3]: Yes, I verified and the test works with 1.17 as well

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 18]
```diff
@@ -0,0 +1,683 @@
...
+	for _, val := range policytel {
+		if val != policy {
+			t.Errorf("Apply policy mismatch: got %v, want %v", policytel, policy)
+		}
+	}
+}
+
+// configureOTG configures the interfaces and BGP protocols on an OTG, including advertising some
+// (faked) networks over BGP.
+func configureOTG(t *testing.T, otg *ondatra.OTG, expectedRoutes int32) (gosnappi.Config, otgutils.ExpectedState) {
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:340]: Looking at the pairwise diff between the ATE and OTG version, I can't help but notice that the OTG version is longer, partly because it wants a name assigned for every node in the config tree. Can some of these names be omitted?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 19]
```diff
@@ -0,0 +1,683 @@
...
+		SetAsType(gosnappi.BgpV4PeerAsType.EBGP)
+	dstBgp6Peer := dstBgp.Ipv6Interfaces().Add().
+		SetIpv6Name(dstIpv6.Name()).
+		Peers().Add().
+		SetName(dstBgp6Name).
+		SetPeerAddress(dstIpv6.Gateway()).
+		SetAsNumber(ateAS).
+		SetAsType(gosnappi.BgpV6PeerAsType.EBGP)
+
+	prefixInt4, _ := strconv.Atoi(strings.Split(advertisedRoutesv4CIDR, "/")[1])
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:416]: Define separate constants for the prefixlen and the IP netblock? We want to avoid reprocessing of test parameters because that's error prone.

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 20]
```diff
@@ -0,0 +1,683 @@
...
+	expected := otgutils.ExpectedState{
+		Bgp4: map[string]otgutils.ExpectedBgpMetrics{
+			srcBgp4Peer.Name(): {Advertised: 0, Received: expectedRoutes},
+			dstBgp4Peer.Name(): {Advertised: routeCount, Received: 0},
+		},
+		Bgp6: map[string]otgutils.ExpectedBgpMetrics{
+			srcBgp6Peer.Name(): {Advertised: 0, Received: expectedRoutes},
+			dstBgp6Peer.Name(): {Advertised: routeCount, Received: 0},
+		},
+		Flow: map[string]otgutils.ExpectedFlowMetrics{
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:476]: Is the flow state used anywhere?

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 21]
```diff
@@ -0,0 +1,685 @@
...
+	t.Logf("BGP v6 last est time :%d", lestTimev6)
+	if lestTime < lastFlapTime {
+		t.Errorf("Bad last-established timestamp: got %v, want >= %v", lestTime, lastFlapTime)
+	}
+
+	// Check BGP Transitions
+	estTrans := nbr.GetEstablishedTransitions()
+	t.Logf("Got established transitions: %d", estTrans)
+	if estTrans != 1 {
+		// t.Errorf("Wrong established-transitions: got %v, want 1", estTrans)
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:264]: This is something I wanted to discuss...but basically every time I run the test, that error is always triggered as the established transitions is always 0.

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 18 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:340]: Unfortunately setting the name is mandatory in the model for every object/node. Maybe the model could be improved by adding some deafult values to names, mac addresses (see the previous comments)...etc

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 19 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:416]: Done

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 18 above)
Comment (liulk) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:340]: I just realized that telemetry will need the node names.

Since node names are mandatory, how about making it more streamlined?

```
config := otg.NewConfig(t)
srcDev := config.GetOrCreateDevice("src") // or just config.Device()
srcPort := srcDev.GetOrCreatePort("port1") // or just srcDev.Port()
srcEth := srcPort.GetOrCreateEthernet("eth1").SetMAC(...)
srcIPv4 := srcEth.GetOrCreateIPv4("eth1:ipv4").SetAddress(...).SetGateway(...)
```

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 22]
```diff
@@ -0,0 +1,683 @@
...
+	config := otg.NewConfig(t)
+	srcPort := config.Ports().Add().SetName("port1")
+	dstPort := config.Ports().Add().SetName("port2")
+
+	srcDev := config.Devices().Add().SetName(ateSrc.Name)
+	srcEth := srcDev.Ethernets().Add().
+		SetName(ateSrc.Name + ".eth").
+		SetPortName(srcPort.Name()).
+		SetMac(ateSrc.MAC)
+	srcIpv4 := srcEth.Ipv4Addresses().Add().
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:351]: By the way, we would like to follow Go's rule about initialism in identifiers.

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Rename Ipv4 to IPv4, Ipv6 to IPv6, etc?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 20 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:476]: Removed

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 23]
```diff
@@ -0,0 +1,683 @@
...
+	flowipv6.Size().SetFixed(512)
+	flowipv6.Rate().SetPps(trafficPacketRate)
+	flowipv6.Duration().SetChoice("continuous")
+	e2 := flowipv6.Packet().Add().Ethernet()
+	e2.Src().SetValue(srcEth.Mac())
+	v6 := flowipv6.Packet().Add().Ipv6()
+	v6.Src().SetValue(srcIpv6.Address())
+	v6.Dst().Increment().SetStart(strings.Split(advertisedRoutesv6CIDR, "/")[0]).SetCount(routeCount)
+
+	expected := otgutils.ExpectedState{
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:467]: - otgutils.ExpectedState includes advertised and received routes and these values are used by the functions we have in validations.go (.e.g AllBgp4Up, AllBGP6Up). 
- So they are checking for the state of the peers and also the route counts. 
- The 4 sub cases from TestBGPPolicy have different route requirements and that is why I'm passing to the configureOTG function
- None of this is done in the ATE test. If you prefer, I can remove these checks for the OTG

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 22 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:351]: I have replaced all occurrences with proper initialism. (e.g. BGP, IPv4 ... )

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 18 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:340]: I made some changes...see below, but for a function like "GetOrCreateDevice" we would need to create a new wrapper as gosnappi doesn't support it.
```go
	srcDev := config.Devices().Add().SetName(ateSrc.Name)
	srcEth := srcDev.Ethernets().Add().SetName(ateSrc.Name + ".Eth")
	srcEth.SetPortName(srcPort.Name()).SetMac(ateSrc.MAC)
	srcIpv4 := srcEth.Ipv4Addresses().Add().SetName(ateSrc.Name + ".IPv4")
	srcIpv4.SetAddress(ateSrc.IPv4).SetGateway(dutSrc.IPv4).SetPrefix(int32(ateSrc.IPv4Len))
```

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (liulk) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:419]: I noticed that RT-1.2 and RT-1.3 test plan has the language "Advertise prefixes from ATE port-1, observe received prefixes at ATE port-2." but other BGP tests just want to check that the session is up.

I think just checking the session is up should just use Await, and that does not need a separate helper. On the other hand, I can see that some helper may be useful to check the AFT telemetry that the advertised routes are present. I think we can probably share the same code between DUT, ATE and OTG if it's only using the Ondatra Telemetry API. Such code could go under a separate bgputils package.

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 24]
```diff
@@ -0,0 +1,629 @@
...
+
+	t.Logf("Pushing config to ATE and starting protocols...")
+	otg.PushConfig(t, config)
+	otg.StartProtocols(t)
+	return config, expected
+}
+
+// verifyTraffic confirms that every traffic flow has the expected amount of loss (0% or 100%
+// depending on wantLoss, +- 2%)
+func verifyTraffic(t *testing.T, ate *ondatra.ATEDevice, c gosnappi.Config, wantLoss bool) {
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:438]: Maybe there could be a flowutils package that checks the flow telemetry alone, that could be shared between ATE and OTG? otgutils.PrintFlowMetrics would be inlined to flowutils.VerifyTraffic().

The idea is that it just needs a list of flow names are arguments, and use (*ondatra.Device).Telemetry().Flow(flowName) to access the telemetry.

Comment (liulk) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:438]: Actually the util function only needs to check one flow at a time, and rather than asserting whether the traffic happened on behalf of the test, it should return whether the traffic passed as a bool.

```
package flowutils

func HasTraffic(t testing.TB, device *ondatra.Device, flowName string) bool {
  ...
}
```

We are moving away from the so-called assertion helpers. We can still have helpers that return the result of a check, but leave it up to the caller to decide what to do with that result.

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 25]
```diff
@@ -0,0 +1,629 @@
...
+	otg := ate.OTG()
+	otgutils.PrintFlowMetrics(t, otg, c)
+	for _, f := range c.Flows().Items() {
+		t.Logf("Verifying flow metrics for flow %s\n", f.Name())
+		recvMetric := otg.Telemetry().Flow(f.Name()).Get(t)
+		txPackets := recvMetric.GetCounters().GetOutPkts()
+		rxPackets := recvMetric.GetCounters().GetInPkts()
+		lostPackets := txPackets - rxPackets
+		lossPct := lostPackets * 100 / txPackets
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:447]: In OTG model we don't have the LossPct stat, so we can only calculate it based on the Tx/Rx packets stats

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 24 above)
Comment (liulk) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:438]: Okay, so if ATE and OTG telemetry are significantly different, then utils like HasTraffic can go into otgutils. But so far I've only seen that OTG is missing LossPct(), and haven't seen any other differences that could affect HasTraffic. I think for helpers that aren't affected by the ATE and OTG telemetry differences, it would probably be best to make them work for both and avoid ATE-ism or OTG-ism.

I still think the helper should avoid having the for loop over all flows. The helper should check one flow at a time, and the caller will call it using explicit flow names.

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:419]: I have removed the extra route check for BGP and created a local function for BGP sessions check using Watch and Await ondatra telemetry utilities. I've also removed the BGP related functions from the otgutils package. 
The otg bgp verification is required in order to ensure successfully that traffic is started. Unlike the ATE, in OTG the "StartProtocols" is asynchronous

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 24 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:428]: According to the latest reviews on PR#259 now we have only print utilities under otgutils package. In that context may we request your approval on this PR#240.

**Comment by SubhajitPalKeysight** [comment]
@bstoll , @liulk  - Would you please comment if any any further open / pending items on this PR. 
Thanks
subhajit

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 26]
```diff
@@ -120,6 +128,7 @@ var (
 
 // configureDUT configures all the interfaces on the DUT.
 func configureDUT(t *testing.T, dut *ondatra.DUTDevice) {
+	deviations.InterfaceEnabled = ygot.Bool(true)
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:131]: Please revert this change?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 27]
```diff
@@ -149,7 +158,7 @@ func bgpCreateNbr(localAs, peerAs uint32, policy string) *telemetry.NetworkInsta
 	nbrs := []*bgpNeighbor{nbr1v4, nbr2v4, nbr1v6, nbr2v6}
 
 	d := &telemetry.Device{}
-	ni1 := d.GetOrCreateNetworkInstance(*deviations.DefaultNetworkInstance)
+	ni1 := d.GetOrCreateNetworkInstance("default")
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:161]: Please use *deviations.DefaultNetworkInstance throughout?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 28]
```diff
@@ -325,123 +334,168 @@ func verifyPolicyTelemetry(t *testing.T, dut *ondatra.DUTDevice, policy string)
...
 	t.Logf("Starting traffic")
-	ate.Traffic().Start(t, allFlows...)
+	otg.StartTraffic(t)
 	time.Sleep(trafficDuration)
-	ate.Traffic().Stop(t)
 	t.Logf("Stop traffic")
+	otg.StopTraffic(t)
+}
+
+func verifyOtgBgpTelemetry(t *testing.T, otg *otg.OTG, c gosnappi.Config, ipType, state string) {
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:466]: verifyOtgBgpTelemetry -> verifyOTGBGPTelemetry because of golang initialism?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 29]
```diff
@@ -325,123 +334,168 @@ func verifyPolicyTelemetry(t *testing.T, dut *ondatra.DUTDevice, policy string)
...
+	for _, d := range c.Devices().Items() {
+		switch ipType {
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:468]: I noticed that verifyOTGBGPTelemetry is always called in pairs of "IPv4" and "IPv6" anyways, so maybe we can just remove the switch?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 30]
```diff
@@ -325,123 +334,168 @@ func verifyPolicyTelemetry(t *testing.T, dut *ondatra.DUTDevice, policy string)
...
+		case "IPv4":
+			for _, ip := range d.Bgp().Ipv4Interfaces().Items() {
+				for _, configPeer := range ip.Peers().Items() {
+					nbrPath := otg.Telemetry().BgpPeer(configPeer.Name())
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:472]: Are lines 472-480 and 486-494 the same code that can be factored to a common function? Or does the "ip" variable have different types for IPv4 and IPv6?

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 31]
```diff
@@ -325,123 +334,168 @@ func verifyPolicyTelemetry(t *testing.T, dut *ondatra.DUTDevice, policy string)
...
+					_, ok := nbrPath.SessionState().Watch(t, time.Minute,
+						func(val *otgtelemetry.QualifiedE_BgpPeer_SessionState) bool {
+							return val.IsPresent() && val.Val(t).String() == state
+						}).Await(t)
+					if !ok {
+						fptest.LogYgot(t, "BGP reported state", nbrPath, nbrPath.Get(t))
+						t.Fatal("No BGP neighbor formed")
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:479]: This is not really a fatal. Use `t.Error` instead?

Also, log which peer with the error?

```
t.Errorf("No BGP neighbor formed for peer %s", configPeer.Name())
```

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go [hunk 32]
```diff
@@ -461,36 +515,35 @@ func TestEstablish(t *testing.T) {
...
 
-	t.Log("Configure Network Instance")
-
-	dutConfNIPath := dut.Config().NetworkInstance(*deviations.DefaultNetworkInstance)
-	dutConfNIPath.RouterId().Replace(t, dutDst.IPv4)
-
 	// Configure BGP+Neighbors on the DUT
 	t.Logf("Start DUT BGP Config")
-	dutConfPath := dut.Config().NetworkInstance(*deviations.DefaultNetworkInstance).Protocol(telemetry.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
+	dutConfPath := dut.Config().NetworkInstance("default").Protocol(telemetry.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
```
Comment (liulk) [inline feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:520]: Please restore the DUT config changes. In particular, "DUT BGP Config before" may fatal because devices had no initial BGP config, and the default network instance should not be replaced at the instance level because it may break other things.

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 28 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:466]: Done

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 27 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:161]: Done

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 26 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:131]: Done

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 32 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:520]: Brought in the changes from the ate test

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 30 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:472]: The items in the for loop are different for each ip type, but I simplified the function a little bit

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 29 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:468]: Done

File: feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go (hunk 31 above)
Comment (octpetre) [reply feature/bgp/policybase/otg_tests/route_installation_test/route_installation_test.go:479]: Done

**Review by liulk — APPROVED** [review]
/gcbrun
